File: lib/cards.ex
defmodule Cards do
  @moduledoc """
  Documentation for Cards Module

  Shuffle function is taken from "Etudes for Elixir", J. David Eisenberg:
  https://github.com/oreillymedia/etudes-for-elixir/blob/master/code/ch09-01/cards.ex
  """

  require Card

  @doc """
  Shuffle a list into random order using the Fisher-Yates method.
  From "Etudes for Elixir"
  """

  @spec shuffle([Card.t()]) :: [Card.t()]
  def shuffle(list) do
    :random.seed(:erlang.now())
    shuffle(list, [])
  end

  # From "Etudes for Elixir":
  #
  # The helper function takes a list to shuffle as its first
  # argument and the accumulated (shuffled) list as its second
  # argument.

  # When there are no more cards left to shuffle,
  # return the accumulated list.

  def shuffle([], acc) do
    acc
  end

  # From "Etudes for Elixir":
  #
  # This is easier to understand if you look at it as a
  # physical process. Split the deck at a random point.
  # Put the part above the "split point" aside (leading), and
  # take the first card (h) off the part below the split (t).
  # That first card goes onto a new pile ([h | acc]).
  # Now put together the part above the split and the
  # part below the split (leading ++ t) and go through
  # the process with the deck (which now has one less card).
  # This keeps going until you run out of cards to shuffle;
  # at that point, all the cards will have gotten to the
  # new pile, and that's your shuffled deck.

  def shuffle(list, acc) do
    {leading, [h | t]} = Enum.split(list, :random.uniform(Enum.count(list)) - 1)
    shuffle(leading ++ t, [h | acc])
  end

  @doc """
  Create a deck of 52 cards in the form [{Card, "Suit::Club", "Value::Three"},
  {Card, "Suit::Diamond", "Value::Ace"}...]
  """
  @spec make_deck() :: [Card.t()]
  def make_deck() do
    for suit <- [
          "Suit::Club",
          "Suit::Diamond",
          "Suit::Heart",
          "Suit::Spade"
        ] do
      for value <- [
            "Value::Two",
            "Value::Three",
            "Value::Four",
            "Value::Five",
            "Value::Six",
            "Value::Seven",
            "Value::Eight",
            "Value::Nine",
            "Value::Ten",
            "Value::Jack",
            "Value::Queen",
            "Value::King",
            "Value::Ace"
          ] do
        Card.record(suit: suit, value: value)
      end
    end
    |> List.flatten()
  end

  # Compare cards
  # First let's assign them values

  @doc """
  value(card) -> non_neg_integer

  ### Examples

  iex(3)> Cards.value({Card, "Suit::Spade", "Value::Ace"})
  14

  """

  @spec value(Card.t()) :: non_neg_integer
  def value(card) do
    case card do
      {_, _, value} ->
        case value do
          "Value::Two" -> 2
          "Value::Three" -> 3
          "Value::Four" -> 4
          "Value::Five" -> 5
          "Value::Six" -> 6
          "Value::Seven" -> 7
          "Value::Eight" -> 8
          "Value::Nine" -> 9
          "Value::Ten" -> 10
          "Value::Jack" -> 11
          "Value::Queen" -> 12
          "Value::King" -> 13
          "Value::Ace" -> 14
          _ -> 0
        end
    end
  end

  @doc """
  first_is_less(card1, card1)
  true if card1 < card2
  false otherwise

  ### Examples

  iex(7)> Cards.first_is_less({Card, "Suit::Spade", "Value::Two"}, {Card, "Suit::Heart", "Value::Queen"})
  false

  """

  @spec first_is_less(Card.t, Card.t) :: boolean
  def first_is_less(card1, card2) do
    value(card1) < value(card2)
  end

  @doc """
  cards_are_equal(card1, card2)
  true if card1 == card2
  false otherwise

  ### Examples

  iex(8)> Cards.cards_are_equal({Card, "Suit::Spade", "Value::Ace"}, {Card, "Suit::Heart", "Value::Queen"})
  false

  """

  @spec cards_are_equal(Card.t, Card.t) :: boolean
  def cards_are_equal(card1, card2) do
    value(card1) == value(card2)
  end
end

File: nif_serde_rustler/native/cardsnif/src/cards.rs
// #![allow(dead_code)]

extern crate rand;
use serde::{Deserialize, Serialize};

#[derive(Debug, Copy, Clone, PartialEq, Serialize, Deserialize)]
pub enum Suit {
    #[serde(rename = "Suit::Club")]
    Club,
    #[serde(rename = "Suit::Diamond")]
    Diamond,
    #[serde(rename = "Suit::Heart")]
    Heart,
    #[serde(rename = "Suit::Spade")]
    Spade,
}

type Suits = [Suit; 4];

pub fn get_suits() -> Suits {
    [Suit::Club, Suit::Diamond, Suit::Heart, Suit::Spade]
}

#[derive(Debug, Copy, Clone, PartialEq, Serialize, Deserialize)]
pub enum Value {
    #[serde(rename = "Value::Two")]
    Two,
    #[serde(rename = "Value::Three")]
    Three,
    #[serde(rename = "Value::Four")]
    Four,
    #[serde(rename = "Value::Five")]
    Five,
    #[serde(rename = "Value::Six")]
    Six,
    #[serde(rename = "Value::Seven")]
    Seven,
    #[serde(rename = "Value::Eight")]
    Eight,
    #[serde(rename = "Value::Nine")]
    Nine,
    #[serde(rename = "Value::Ten")]
    Ten,
    #[serde(rename = "Value::Jack")]
    Jack,
    #[serde(rename = "Value::Queen")]
    Queen,
    #[serde(rename = "Value::King")]
    King,
    #[serde(rename = "Value::Ace")]
    Ace,
}

type Values = [Value; 13];

pub fn get_values() -> Values {
    [
        Value::Two,
        Value::Three,
        Value::Four,
        Value::Five,
        Value::Six,
        Value::Seven,
        Value::Eight,
        Value::Nine,
        Value::Ten,
        Value::Jack,
        Value::Queen,
        Value::King,
        Value::Ace,
    ]
}

#[derive(Debug, Copy, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename = "Elixir.Card")]
pub struct Card(pub Suit, pub Value);

impl Card {
    pub fn value(&self) -> u32 {
        match self.1 {
            Value::Two => 2,
            Value::Three => 3,
            Value::Four => 4,
            Value::Five => 5,
            Value::Six => 6,
            Value::Seven => 7,
            Value::Eight => 8,
            Value::Nine => 9,
            Value::Ten => 10,
            Value::Jack => 11,
            Value::Queen => 12,
            Value::King => 13,
            Value::Ace => 14,
        }
    }
}

pub fn first_is_less(first: &Card, second: &Card) -> bool {
    first.value() < second.value()
}

#[cfg(test)]
mod first_is_less {
    use super::*;

    #[test]
    fn two_less_than_ace() {
        let first = Card(Suit::Club, Value::Two);
        let second = Card(Suit::Heart, Value::Ace);
        assert_eq!(first_is_less(&first, &second), true);
    }

    #[test]
    fn jack_not_less_than_nine() {
        let first = Card(Suit::Diamond, Value::Jack);
        let second = Card(Suit::Spade, Value::Nine);
        assert_eq!(first_is_less(&first, &second), false);
    }
}

pub fn cards_are_equal(first: &Card, second: &Card) -> bool {
    first.value() == second.value()
}

#[cfg(test)]
mod cards_are_equal {
    use super::*;

    #[test]
    fn spade_ace_eq_diamond_ace() {
        let spade_ace = Card(Suit::Spade, Value::Ace);
        let diamond_ace = Card(Suit::Diamond, Value::Ace);
        assert_eq!(cards_are_equal(&spade_ace, &diamond_ace), true);
    }
}

fn make_deck() -> [Card; 52] {
    let suits = get_suits();
    let values = get_values();
    let mut deck: [Card; 52] = [Card(Suit::Spade, Value::Ace); 52];
    let mut i = 0;
    for suit in suits.iter() {
        for value in values.iter() {
            deck[i] = Card(*suit, *value);
            i += 1;
        }
    }
    deck
}

#[cfg(test)]
mod make_deck {
    use super::*;

    #[test]
    fn len_of_deck_is_52() {
        let deck = make_deck();
        assert_eq!(deck.len(), 52);
    }

    #[test]
    fn check_first_4_cards() {
        let deck = make_deck();
        let four_cards = [
            Card(Suit::Club, Value::Two),
            Card(Suit::Club, Value::Three),
            Card(Suit::Club, Value::Four),
            Card(Suit::Club, Value::Five),
        ];
        assert_eq!(deck[0..4], four_cards[..]);
    }
}

pub fn shuffled_deck() -> [Card; 52] {
    use cards::rand::prelude::*;
    let mut deck = make_deck();
    let len = deck.len();
    // println!("len: {}", len);
    for i in 0..(len - 2) {
        let ind = rand::thread_rng().gen_range(i + 1, len);
        // println!("i: {}, ind: {}",i , ind);
        deck.swap(i, ind);
    }
    deck
}

#[cfg(test)]
mod shuffled_deck {
    use super::*;
    use cards::rand::prelude::*;

    #[test]
    fn print_rand() {
        //let x: u8 = random();
        let x: u8 = rand::thread_rng().gen_range(52, 53);
        // let _shuffled_deck = shuffled_deck();
        // let mut _i = 1;
        // for card in _shuffled_deck.iter() {
        //     println!("{}: {:?}", _i, card);
        //     _i += 1;
        // }
        assert_eq!(x, 52);
    }
}
